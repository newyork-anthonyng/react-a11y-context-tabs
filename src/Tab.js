import React from "react";
import PropTypes from "prop-types";
import callAll from "./callAll";

class Tab extends React.Component {
  componentDidMount() {
    this.context.onTabMount(this.props.id, this.handleFocus);
  }

  handleFocus = () => {
    this.node.focus();
  };

  isSelected = () => {
    return this.props.id === this.context.selectedId;
  };

  handleClick = () => {
    this.context.onActivate(this.props.id);
  };

  handleRef = node => {
    this.node = node;

    const { innerRef } = this.props;
    if (typeof innerRef === "function") innerRef(node);
  };

  render() {
    const { innerRef, className, children, id, onClick, ...rest } = this.props;
    const { selectedTabClassName } = this.context;
    const isSelected = this.isSelected();

    let updatedClassName = `${className}`;
    if (isSelected) updatedClassName += ` ${selectedTabClassName}`;

    return (
      <div
        className={updatedClassName}
        aria-selected={isSelected ? "true" : "false"}
        id={`${id}--tab`}
        aria-controls={`${id}--panel`}
        tabIndex={isSelected ? 0 : "-1"}
        onClick={callAll(onClick, this.handleClick)}
        ref={this.handleRef}
        {...rest}
      >
        {children}
      </div>
    );
  }
}

Tab.contextTypes = {
  onActivate: PropTypes.func,
  onTabMount: PropTypes.func,
  onTabUnmount: PropTypes.func,
  selectedId: PropTypes.string,
  selectedTabClassName: PropTypes.string
};

Tab.propTypes = {
  id: PropTypes.string.isRequired,
  innerRef: PropTypes.func,
  className: PropTypes.string,
  children: PropTypes.node,
  onClick: PropTypes.func
};

Tab.defaultProps = {
  className: "react-tabs__tab"
};

export default Tab;
