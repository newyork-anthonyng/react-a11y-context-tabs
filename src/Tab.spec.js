import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Tab from "./Tab";

const defaultProps = {
  id: "foo"
};

const defaultContext = {
  onActivate: () => {},
  onTabMount: () => {},
  selectedId: "bar",
  selectedTabClassName: "tab--selected"
};

it("should call onTabMount when mounting", () => {
  const cb = jest.fn();
  const wrapper = mount(
    <Tab {...defaultProps}>
      <h1>Hello World</h1>
    </Tab>,
    {
      context: { ...defaultContext, onTabMount: cb }
    }
  );

  expect(cb).toHaveBeenCalledTimes(1);
});

it("should call onTabMount with handleFocus function", () => {
  const cb = jest.fn();
  const wrapper = mount(
    <Tab {...defaultProps}>
      <h1>Hello World</h1>
    </Tab>,
    {
      context: { ...defaultContext, onTabMount: cb }
    }
  );
  const tabNode = wrapper.instance().node;
  tabNode.focus = jest.fn();

  cb.mock.calls[0][1]();

  expect(tabNode.focus).toHaveBeenCalledTimes(1);
});

it("should render correctly when selected", () => {
  const wrapper = mount(
    <Tab {...defaultProps}>
      <h1>Hello World</h1>
    </Tab>,
    {
      context: { ...defaultContext, selectedId: "foo" }
    }
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should render correctly when not selected", () => {
  const wrapper = mount(
    <Tab {...defaultProps}>
      <h1>Hello World</h1>
    </Tab>,
    {
      context: defaultContext
    }
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should call onActivate context function when clicking on Tab", () => {
  const cb = jest.fn();
  const wrapper = mount(
    <Tab {...defaultProps}>
      <h1>Hello World</h1>
    </Tab>,
    {
      context: { ...defaultContext, onActivate: cb }
    }
  );

  const divEl = wrapper.find("div");
  divEl.simulate("click");

  expect(cb).toHaveBeenCalledTimes(1);
});

it("should call custom onClick callback when clicking on Tab", () => {
  const cb = jest.fn();
  const wrapper = mount(
    <Tab {...defaultProps} onClick={cb}>
      <h1>Hello World</h1>
    </Tab>,
    {
      context: defaultContext
    }
  );

  const divEl = wrapper.find("div");
  divEl.simulate("click");

  expect(cb).toHaveBeenCalledTimes(1);
});

it("should run innerRef prop", () => {
  const cb = jest.fn();
  const wrapper = mount(
    <Tab {...defaultProps} onClick={cb} innerRef={cb}>
      <h1>Hello World</h1>
    </Tab>,
    {
      context: defaultContext
    }
  );

  expect(cb).toHaveBeenCalledTimes(1);
  expect(cb.mock.calls[0]).toMatchSnapshot();
});
