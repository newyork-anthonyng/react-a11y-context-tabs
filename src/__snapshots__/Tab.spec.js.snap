// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when not selected 1`] = `
<Tab
  className="react-tabs__tab"
  id="foo"
>
  <div
    aria-controls="foo--panel"
    aria-selected="false"
    className="react-tabs__tab"
    id="foo--tab"
    onClick={[Function]}
    tabIndex="-1"
  >
    <h1>
      Hello World
    </h1>
  </div>
</Tab>
`;

exports[`should render correctly when selected 1`] = `
<Tab
  className="react-tabs__tab"
  id="foo"
>
  <div
    aria-controls="foo--panel"
    aria-selected="true"
    className="react-tabs__tab tab--selected"
    id="foo--tab"
    onClick={[Function]}
    tabIndex={0}
  >
    <h1>
      Hello World
    </h1>
  </div>
</Tab>
`;

exports[`should run innerRef prop 1`] = `
Array [
  <div
    aria-controls="foo--panel"
    aria-selected="false"
    class="react-tabs__tab"
    id="foo--tab"
    tabindex="-1"
  >
    <h1>
      Hello World
    </h1>
  </div>,
]
`;
